### DISCUSSION TITLE
RuntimeError: Failed to authenticate through CAM. HTTP response does not contain 'cam_passport' cookie

### DISCUSSION THREAD

**Discussion by ommore1523** [cubewise-code/tm1py#772] [Q&A]
Hi , I am facing issue with connecting tm1 server 

```text
Versions:
tm1py = 1.7.0
```

#### NOTE:   single sign on is enabled for IBM Cognos Analytics  and Planning Analytics 

```python 
 from TM1py.Services import TM1Service
 with TM1Service(address="<IP_ADDRESS>", port=<PORT> , ssl=False, namespace="<NAMESPACE>", gateway="https://DOMAIN:449/ibmcognos/bi/v1/disp", user="<USERNAME>", password="<PASSWORD>") as tm1:
    print(tm1.server.get_product_version())
```

**Comment by MariusWirtz** [comment]
If you want to authenticate with USER, PASSWORD and NAMESPACE, don't pass the `gateway` parameter.

You can read about the different authentication modes here:
https://code.cubewise.com/tm1py-help-content/how-to-authenticate-to-tm1-with-tm1py

by the way, you probably want to upgrade to TM1py 1.9: `pip install tm1py --upgrade`

**Reply by ommore1523** [reply]
Hi @MariusWirtz   

```python 
 from TM1py.Services import TM1Service
 with TM1Service(address="<IP_ADDRESS>", port=<PORT> , ssl=False, namespace="<NAMESPACE>", user="<USERNAME>", password="<PASSWORD>") as tm1:
    print(tm1.server.get_product_version())
```

I tried to connect with   tm1py  version  1.9.1   and removing gateway .. 
### It stuck in between...
Not get any response... 

 and @MariusWirtz     Should  I ask to check the single sign on configuration check to administrator side ...

**Comment by MariusWirtz** [comment]
Please post the stack trace and error message you get.

Yes. It definitely makes sense to speak to the person that administrates the TM1 Environment.
Make sure that as `port` you pass the HTTPPortNumber that is configured in the tm1s.cfg

**Comment by ommore1523** [comment]
@MariusWirtz as per doucmentation reference I kept gateway instead of username and password.. 
 
```python
with TM1Service(address="192.168.XX.XX", port=4XXXX, ssl=True, namespace="XX", gateway="https://dc1XXXX:449/ibmcognos/bi/v1/disp") as tm1:
    print(tm1.server.get_product_version())
```

### Traceback message 
```bash
Traceback (most recent call last):
  File ".\test.py", line 31, in <module>
    with TM1Service(address="192.168.80.30", port=40652, ssl=True, namespace="KM", gateway="https://dc1epmcatst1:449/ibmcognos/bi/v1/disp") as tm1:
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\TM1Service.py", line 15, in __init__
    self._tm1_rest = RestService(**kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 201, in __init__
    self._start_session(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 353, in _start_session
    token = self._build_authorization_token(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 472, in _build_authorization_token
    return RestService._build_authorization_token_cam(user, password, namespace, gateway, verify)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 494, in _build_authorization_token_cam
    raise RuntimeError(
RuntimeError: Failed to authenticate through CAM. HTTP response does not contain 'cam_passport' cookie
```

and 

**_What should I ask to tm1 adminstrator to verify for ?_** 



#### As per my debug information :
```python 
response = requests.get(
    gateway,
    auth=HttpNegotiateAuth(), 
   verify=verify,params={"CAMNamespace": namespace}
)
```
The above code is not returning  **cam_passport** cookie.

**Reply by MariusWirtz** [reply]
Quick question: Does the single sign-on work on other TM1 frond ends, like for instance TM1Web?

**Reply by ommore1523** [reply]
@MariusWirtz  I tried to login the **gateway url**  using  provided **username** and **password** .  
- I could login to url 
- if I inspected the browser cookies the cam_passport is there in cookies .

![Screenshot from 2022-07-28 14-47-10](https://user-images.githubusercontent.com/52170865/181469788-395ffc37-0ba3-4668-a6d6-4c464ba01f2a.png)

**Reply by MariusWirtz** [reply]
I'm not sure I understand. When you open the gateway URL in the browser, does it log you in automatically or do you have to insert user and password to login?

If you had to provide a username and password in the browser, then you will have to do the same for TM1py.
In that case, you can authenticate like this:
```
with TM1Service(address="", port=12354, ssl=True, user="admin", password="apple", namespace="LDAP") as tm1:
    print(tm1.server.get_product_version())
```

If the TM1 single sign-on works in the browser it should work in TM1py like this:

```
with TM1Service(address="", port=12354, ssl=True, namespace="LDAP", gateway="http://web-host:80/ibmcognos/bi/v1/disp") as tm1:
    print(tm1.server.get_product_version())
```

**Reply by ommore1523** [reply]
@MariusWirtz  After hitting  gateway. It not asked for username and password . It directly redirected to home page . 

**I tried  for given both the approaches :**

- For first approach it hangs and not returning anything

- For second approach it will return cam_passport error message . 



I confused around. If it is the adminstrator level issue or  tm1py package level issue...  ?

**Reply by MariusWirtz** [reply]
Just to make sure, can you please open the URL in a different browser and verify that single sign-on works?

For the first approach, if it hangs and doesn't return, please make sure you got the `port`, `address` and `ssl` right. For port, you need to pass the port that is specified in the tm1s.cfg as the `HttpPortNumber`.

to verify if have the right address and port and SSL you can try out this request in the browser. It doesn't require authentication:

`https://<<address>>:<<port>>/api/v1/$metadata`

**Reply by ommore1523** [reply]
@MariusWirtz   URL :  https://192.168.XX.XX:40XXX/api/v1/$metadata 
Returning :   **This site can’t be reached** error message 


With Gateway URL: https://dc1XXX:449/ibmcognos/bi/v1/$metadata
Return :  **forbidden**

**Reply by MariusWirtz** [reply]
> What should I ask to tm1 adminstrator to verify for ?

Please check with the TM1 administrator about whether you are using the wrong parameters in the URL or why the TM1 REST API isn't exposed

**Reply by ommore1523** [reply]
Ok @MariusWirtz . Thank you for your cooperation.

**Reply by ommore1523** [reply]
Hi @MariusWirtz  
as per your suggetions the url : https://<ip>:<port>/api/v1/$metadata  is returning some xml data.

![Screenshot from 2022-08-08 17-08-12](https://user-images.githubusercontent.com/52170865/183409961-23145a6d-1927-460d-90d7-1f1e95d7bbee.png)


But, 
when i try to run : 

```python 
with TM1Service(address="", port=40672, ssl=True, namespace="KMMM", gateway="http://web-host:80/ibmcognos/bi/v1/disp") as tm1:
    print(tm1.server.get_product_version())
```

Error: 

```bash 
Traceback (most recent call last):
  File ".\test.py", line 9, in <module>
    with TM1Service(address="", port=40653, ssl=True, namespace="KM", gateway="https://dc1epmcatst1:449/ibmcognos/bi/v1/disp") as tm1:
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\TM1Service.py", line 15, in __init__
    self._tm1_rest = RestService(**kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 201, in __init__
    self._start_session(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 354, in _start_session
    token = self._build_authorization_token(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 473, in _build_authorization_token
    return RestService._build_authorization_token_cam(user, password, namespace, gateway, verify)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 495, in _build_authorization_token_cam
    raise RuntimeError(
RuntimeError: Failed to authenticate through CAM. HTTP response does not contain 'cam_passport' cookie
```

And IF I RUN:

```python 
with TM1Service(address="192.168.80.20", port=40653, ssl=True, namespace="KMM",  user="abc", password="Pass") as tm1:
    print(tm1.server.get_product_version())
```
Error:

```bash 

Traceback (most recent call last):
  File ".\test.py", line 4, in <module>
    with TM1Service(address="192.168.80.30", port=40653, ssl=True, namespace="KM",  user="abc", password="Pass") as tm1:
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\TM1Service.py", line 15, in __init__
    self._tm1_rest = RestService(**kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 201, in __init__
    self._start_session(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 369, in _start_session
    response = self.GET(url=url, headers=additional_headers)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 80, in wrapper
    self.verify_response(response=response)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 460, in verify_response
    raise TM1pyRestException(response.text,
TM1py.Exceptions.Exceptions.TM1pyRestException: Text: '' - Status Code: 401 - Reason: 'Unauthorized' - Headers: {'Content-Type': 'text/plain', 'Content-Length': '0', 'Connection': 'keep-alive', 'Set-Cookie': 'TM1SessionId=9Xk1Djr6CfUALhlhZYNie2Zi8_I; Path=/api/; HttpOnly; Secure', 'WWW-Authenticate': 'CAMPassport https://DC1EPMCATST1:449/ibmcognos/bi/v1/disp, CAMNamespace'}

```


In both case , errors are different .

**Comment by rclapp** [comment]
#1 seems to suggest that your gateway URL is wrong. It has reached its maximum number of tries

#2 is clearer, the user name, password, namespace combo is invalid.

First I would ensure that gateway is available by trying it in a browser.

Sent from my mobile device

On Aug 8, 2022 7:45 AM, omkar more ***@***.***> wrote:

Hi @MariusWirtz<https://github.com/MariusWirtz>
as per your suggetions the url : https://:/api/v1/$metadata is returning some xml data.

[Screenshot from 2022-08-08 17-08-12]<https://user-images.githubusercontent.com/52170865/183409961-23145a6d-1927-460d-90d7-1f1e95d7bbee.png>

But,
when i try to run :

with TM1Service(address="", port=40652, ssl=True, namespace="KM", gateway="http://web-host:80/ibmcognos/bi/v1/disp") as tm1:
    print(tm1.server.get_product_version())

Error:

  File ".\test.py", line 9, in <module>
    with TM1Service(address="", port=40652, ssl=True, namespace="KM", gateway="http://web-host:80/ibmcognos/bi/v1/disp") as tm1:
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\TM1Service.py", line 15, in __init__
    self._tm1_rest = RestService(**kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 201, in __init__
    self._start_session(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 354, in _start_session
    token = self._build_authorization_token(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 473, in _build_authorization_token
    return RestService._build_authorization_token_cam(user, password, namespace, gateway, verify)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 487, in _build_authorization_token_cam
    response = requests.get(gateway, auth=HttpNegotiateAuth(), verify=verify,
  File "E:\development\TariffModeling\venv1\lib\site-packages\requests\api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\requests\api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\requests\sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\requests\sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\requests\adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='web-host', port=80): Max retries exceeded with url: /ibmcognos/bi/v1/disp?CAMNamespace=KM (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000002E4D6686550>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))

And IF I RUN:

with TM1Service(address="192.168.80.30", port=40653, ssl=True, namespace="KM",  user="ibmcognos", ***@***.***") as tm1:
    print(tm1.server.get_product_version())

Error:

Traceback (most recent call last):
  File ".\test.py", line 4, in <module>
    with TM1Service(address="192.168.80.30", port=40653, ssl=True, namespace="KM",  user="ibmcognos", ***@***.***") as tm1:
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\TM1Service.py", line 15, in __init__
    self._tm1_rest = RestService(**kwargs)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 201, in __init__
    self._start_session(
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 369, in _start_session
    response = self.GET(url=url, headers=additional_headers)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 80, in wrapper
    self.verify_response(response=response)
  File "E:\development\TariffModeling\venv1\lib\site-packages\TM1py\Services\RestService.py", line 460, in verify_response
    raise TM1pyRestException(response.text,
TM1py.Exceptions.Exceptions.TM1pyRestException: Text: '' - Status Code: 401 - Reason: 'Unauthorized' - Headers: {'Content-Type': 'text/plain', 'Content-Length': '0', 'Connection': 'keep-alive', 'Set-Cookie': 'TM1SessionId=9Xk1Djr6CfUALhlhZYNie2Zi8_I; Path=/api/; HttpOnly; Secure', 'WWW-Authenticate': 'CAMPassport https://DC1EPMCATST1:449/ibmcognos/bi/v1/disp, CAMNamespace'}


In both case , errors are different .

-

**Reply by ommore1523** [reply]
@rclapp 

Login is possible through the browser with same username and password.  

Login process is as follows :
 -  Put `http://<dc1--->:449/ibmcognos/bi/v1/disp` url on  browser in incognito window.
 -  When it pop ups window for username and password same as used which using inside my code
 - `Note`:  It does not ask for any namespace . Only in incognito it asks for username and passoword otherwise it  do login using sso.
 -  They provided me namespace something like `KM` not a `LDAP`
 -

**Reply by MariusWirtz** [reply]
Hi @ommore1523,

option B should work in any case:

``` python
with TM1Service(address="192.168.80.20", port=40653, ssl=True, namespace="KMM",  user="abc", password="Pass") as tm1:
    print(tm1.server.get_product_version())
```

If you get all arguments right it should work. Please note that the namespace is case-sensitive.

**Reply by ommore1523** [reply]
unfortunately Its not working as it should.  May  I know which configuration should I ask to check as tm1s.conf side.  Is there anything where we can diagnostic whether it is working or not..
